### PR TITLE
Use Single Responsibility Form Objects instead of Creators.

### PR DESCRIPTION
**Pull request by terrellt**
This also utilizes a dependency injection container as a dependency on an abstraction - now we don't need to churn the controller when outside dependencies change.

### CODE REVIEW

File: app/injectors/term_injector.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+class TermInjector < VocabularyInjector
+  def term_form
+    @term_form ||= term_form_factory.new(built_term, term_repository)
+  end
```
Comment (jechols) [inline app/injectors/term_injector.rb]: I think this is exactly the same as its parent class, and could be removed

**Comment by jechols** [comment]
First, +1 for pulling out the responders, I prefer the more inline approach when the logic is so simple and so specific to the controller.

The injectors I'm not sold on.  They seem to be responsible for a lot of indirection of fairly simple instantiations of isolated classes / logic from elsewhere.  For instance, `all_vocabs_query`.  The controller calls the injector which returns a lambda which calls AllVocabsQuery.  Simply calling AllVocabsQuery in the controller feels okay to me.  The logic is already elsewhere and easily changed without messing with the controller.  I think I feel the same about all the functions I see in the two injectors.  It feels like you've already decoupled all the real logic, and we're not really gaining anything with the extra abstraction.

As a side note, I'd avoid subclassing the injectors.  It doesn't seem like they share more than one or two methods, and it confuses an already complicated situation.

I think the addition of the form objects is a good thing.  I'd prefer to not use SimpleDelegator, and be more explicit about what calls down and what doesn't (via delegate, even, just not via the SimpleDelegator class).

I think I liked the creator, at least conceptually.  But with the form and validators abstracted out, it may be a lot simpler to just have logic in the controller.

I'm not sure about the TermUri class - in cases where the URI isn't exactly the way we do things, it gets weird.  Try it out with `RDF::URI("example.com/vocabulary/rdf/v1/termname")` or even a more real-world case, `RDF::URI("http://www.loc.gov/standards/mods/modsrdf/v1/#locationCopy")`.  We may need that class to be smarter if we have any use cases where the namespace contains extra path elements or the children have `#` in the name.  This class also wasn't present before; was there a bug previously?  It seems like unnecessary complexity if it's not fixing some issue.

**Comment by terrellt** [comment]
> For instance, all_vocabs_query. The controller calls the injector which returns a lambda which calls AllVocabsQuery. Simply calling AllVocabsQuery in the controller feels okay to me.

So normally I would agree with this. However, the real benefit here is abstracting away dependencies. AllVocabsQuery takes a sparql_client as an argument in order to satisfy the open/closed principle. Great! However, if instantiation isn't moved to a single reusable location then should that signature change it needs to be changed in all relevant places, violating OCP in any places where it's used. This way the injector can simply be used in any cases where it's needed. The proc is just a nice shortcut - now the interface is "something that responds to #call", which is really close to the original interface. `Depend on abstractions`

> I'd prefer to not use SimpleDelegator, and be more explicit about what calls down and what doesn't (via delegate, even, just not via the SimpleDelegator class).
Again, normally I'd agree. Unfortunately that would duplicate the property definitions, no? In order for a form to work we would have to A) Add the property to the object model and B) Define that property to delegate. SimpleDelegator here is getting a little fancy, but I see this as decorating validations and form stuff onto an object model - which I think is okay? More discussion would be helpful here.

> I think I liked the creator, at least conceptually. But with the form and validators abstracted out, it may be a lot simpler to just have logic in the controller.

Is this saying that the form objects may not be the way to go?

> This class also wasn't present before; was there a bug previously?

Yes. https://github.com/OregonDigital/ControlledVocabularyManager/pull/76/files#diff-a794b7e8e97fc35e67f3a4749254c721R41 fails without it. The TermUri class was purely to encapsulate obtaining the leaf piece of a term's ID to determine if it's blank or not. For a term, "/vocab/" means there isn't an ID, even though technically its ID is "/vocab/". Maybe it needs tests - I don't think it needs to work outside our use cases, though. Maybe that means we should move it out of lib?

**Comment by terrellt** [comment]
> As a side note, I'd avoid subclassing the injectors. It doesn't seem like they share more than one or two methods, and it confuses an already complicated situation.

Forgot to reply to this. I agree - I was already getting confused about what interfaces I had access to when I wrote it. I'll play with a couple options - I see injector hierarchy and duplication as options.

**Comment by terrellt** [comment]
@jechols Does https://github.com/OregonDigital/ControlledVocabularyManager/commit/c1e71b6d2e06bf00175e3ece9a30e9035519d0bf look better for the injectors?

**Comment by jechols** [comment]
+1 on "Compose injectors".  still looking over stuff.  I think you're right on the form delegation.  I don't like the magick, but if the idea is decoration with validation, it would be a lot of work to explicitly delegate to all the bits Rails needs.

**Comment by jechols** [comment]
+1 on pulling out filters - didn't really pay attention before, but I tend to dislike filters myself

**Comment by jechols** [comment]
I think the term url thing needs some tests for sure - how did I miss that?  If it's specific to our use cases I'd prefer at least a note.  I don't know if it should live elsewhere per se, but it should at least be clear how it's intended to work (and maybe where it doesn't make sense to use)

**Comment by jechols** [comment]
+1 on tests and comments

**Comment by jechols** [comment]
I think I'm good with a merge.  Unless anybody else has thoughts/concerns?

**Pull request merged**